portal-impl/src/com/liferay/portlet/internal/HeaderResponseImpl.java

```diff
@@ -293,20 +292,36 @@ else if (_printWriter != null) {
 	}
 
 	private String _addClosingTags(String xml) {
-		if (xml.contains("<link")) {
-			xml = xml.replaceAll("(<link[^>]*)>(?!\\s*</link>)", "$1></link>");
-		}
+		xml = _addClosingTags(xml, "<link", "</link>");
+		xml = _addClosingTags(xml, "<LINK", "</LINK>");
```

Comment:
@ngriffin7a I am not really getting what is the goal here. Can you describe what logic is needed here? Better to create some unit test to show me the expected behavior. You can make these method static, and use ReflectionTestUtil to invoke them from the test, that way you don't have to make them public or protected.

Reply:
@shuyangzhou The `<link>` and `<meta>` HTML tags are not permitted (by the HTML spec) to have closing `</link>` and `</meta>` tags, respectively. But since we are using an XML parser to parse the markup, it is necessary for the XML parser to work with well-formed XML.

For example, this code would change the following:
```
<link rel="stylesheet" type="text/css" href="foo.css">
<meta name="author" content="Julien">
```
to:
```
<link rel="stylesheet" type="text/css" href="foo.css"></link>
<meta name="author" content="Julien"></link>
```

Is that a good enough explanation? Do you still require the unit tests?